Draft: Home page redesign prototype

just-max opened the pull request:
I've put together a prototype for a start page redesign (#597).

### Screenshots

#### Desktop

![desktop](https://user-images.githubusercontent.com/17551908/182035460-71d00e03-9e68-43d6-9817-78e49e8b4567.png)
<details><summary><h4>Tablet and Mobile</h4></summary>
<p float="left" width="100%">
<img width="72.67588%" src="https://user-images.githubusercontent.com/17551908/182035406-633ac18c-1757-41b5-aab1-41c4127b9f9a.png" />
<img width="26.32411%" src="https://user-images.githubusercontent.com/17551908/182035376-c69fca93-ae92-43e4-b98d-26e691f46a09.png" />
</p>
</details>

### Ideas/Background/Plan

Things that are not yet fully implemented marked with [*].

Design is based on the course overview page.

The idea for the sections (names could be changed, for now it's just what I came up with). Based on the idea of moving the things I want to see when I visit the home page close to the top and accessible in a single click.

- Suggested: Currently live streams, then completed VoDs, sorted by most recent completed first. If the user is logged in, only show streams/VoDs from pinned or registered courses (the ones that currently show up under "My Courses") that are not watched (currently considered "watched" if marked as watched or >50% watch percentage); otherwise from all courses. When a user opens TUM-Live, they are likely to want to join a lecture that is currently streaming or watch a recently completed VoD that they missed (at least that's my thinking). [\*] Probably limit to about two rows of suggestions depending on screen size.
- My Courses: Combine pinned courses and registered courses. These are the courses a user is interested in seeing. Pinned courses are shown first. For each course  ([\*] if there are any streams to show), show the current live stream for the course (if there is one), then the most recent completed VoDs. [\*] Limit to one row/a handful of items per course.
- Other Courses: The user is generally not interested in these courses, so just show the same information that is shown on the current home page.
- [*] A section at the bottom would contain the hidden courses, and be collapsed by default. As suggested in the issue, the handling for this should be unified with that for pinned courses (i.e. also saved server-side).

Courses can be [\*] (un-)hidden/pinned from a dropdown menu:

![image](https://user-images.githubusercontent.com/17551908/182036807-11f01923-ea70-4d27-a7af-15940da3ae7c.png)

This has been made keyboard-navigation-friendly too.

I added friendly dates (e.g. "Streamed 1 hour ago" or "Scheduled tomorrow") with [Luxon](https://github.com/moment/luxon). Personally I think this adds some polish, and the exact date/time is shown by hovering (i.e. on the `title`). The current home page kind of does this for upcoming streams. [*] This should be made dynamic, which should be doable since it's contained in a component, but I haven't thought about how to do that best.

I've also moved a whole bunch of components into reusable templates. Not sure if it could be done better with the template engine, maybe someone has better ideas.

### TODO/Next Steps

- Just search for "TODO" on the diff :)
- Didn't add any database queries yet, the things I needed are currently just hacked together.
- Logged-out view could use work.
- Tests

just-max committed:
Home page redesign prototype

Very much work in progress.

MatthiasReumann commented:
Looks neat 💯 

Some more ideas: 
1) I think it would be pretty cool to have thumbnails above the titles of VODs. 
2) We could move `Other Courses` to the left side (Similiar to YouTube and Twitch) so that all the important stuff is in the center.

just-max committed:
Add thumbnails, other design changes.

- Display watch percentage
- Reduce the height of the header
- Always show dropdown button on touchscreens

just-max commented:
Added thumbnails, at least in the UI :) Not sure what to actually show as a thumbnail though

![image](https://user-images.githubusercontent.com/17551908/182362443-b9356c5e-0b16-4eaf-ad76-4ec68891a82a.png)
![image](https://user-images.githubusercontent.com/17551908/182362693-ffd4b25d-38f2-40fb-abf4-669cf9d0d5f5.png)

MatthiasReumann commented:
> Added thumbnails, at least in the UI :) Not sure what to actually show as a thumbnail though
> 
> ![image](https://user-images.githubusercontent.com/17551908/182374947-81e51995-8e0f-47e4-a628-2ac22462c791.png) ![image](https://user-images.githubusercontent.com/17551908/182374665-2dcac993-7031-4c2d-a070-02811b4cf064.png)

How about some kittens? Jokes aside, we already generate thumbnail sprites (See: #472) - maybe we could utilize that for that use-case also.

joschahenningsen commented:
@just-max you have continued some amazing stuff lately, do you mind if I add you to our dev channel on matrix.in.tum.de?

just-max commented:
Sure, I've sent you an email :>

alexanderstephan commented:
I think we could use the existing thumbnails here, but we should probably wait for https://github.com/joschahenningsen/TUM-Live/pull/582 as only a small portion of the videos has thumbnails (> July 22). I am rather busy this week, so we could also show a placeholder until then.

just-max committed:
add cat images

(also make LATEST panel use full width, previously blocked by dropdown menu button)

just-max commented:
currently focused on exams, so I'll probably make (actual) progress on this again next week

for now, maybe we could come back to the kitten idea :)

![image](https://user-images.githubusercontent.com/17551908/183435238-ad6f5e67-12a9-4eee-a1a1-45310ab191b2.png)

just-max committed:
limit number of displayed videos client-side

just-max committed:
add implementation of dynamic timestamp (kind of a hack though)

just-max committed:
move hidden courses server-side

alexanderstephan commented:
Ah, that's fine actually. I am not that good with UI and we've decided that we have a feature freeze for some time until all bugs in are fixed for the new semester.

MatthiasReumann commented:
Hey @just-max! Would you mind if I take over the task of the start page redesign?

just-max commented:
Yeah, go ahead! The latest version of everything is in my branch, if you want to reuse anything, but otherwise I haven't found the motivation to keep working on it, sorry about that.

MatthiasReumann commented:
> Yeah, go ahead! The latest version of everything is in my branch, if you want to reuse anything, but otherwise I haven't found the motivation to keep working on it, sorry about that.

Great. Anyways, thanks for the work!

MatthiasReumann commented:
Closed in favor of #964
